rs/replicated_state/src/replicated_state.rs

```diff
@@ -968,6 +968,67 @@ impl ReplicatedState {
         (expired_callback_count, errors)
     }
 
+    /// Enforces the best-effort message limit by repeatedly shedding the largest
+    /// best-effort messages of the canisters with the largest best-effort memory
+    /// usage until the memory usage drops below the limit.
+    ///
+    /// Time complexity: `O(n * log(n))`.
+    pub fn enforce_best_effort_message_limit(&mut self, limit: NumBytes) {
+        const ZERO: NumBytes = NumBytes::new(0);
+
+        // Check if we need to do anything at all before constructing a priority queue.
+        let mut memory_usage = self
+            .canister_states
+            .values()
+            .fold(NumBytes::new(0), |acc, canister| {
+                acc + canister.system_state.best_effort_message_memory_usage()
+            });
+        if memory_usage <= limit {
+            // No need to do anything.
+            return;
+        }
+
+        // Construct a priority queue of canisters by best-effort message memory usage.
+        let mut priority_queue: BTreeSet<_> = self
+            .canister_states
+            .iter()
+            .filter_map(|(canister_id, canister)| {
+                let memory_usage = canister.system_state.best_effort_message_memory_usage();
+                if memory_usage > ZERO {
+                    Some((memory_usage, *canister_id))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Shed messages from the canisters with the largest memory usage until we are
+        // below the limit.
+        while memory_usage > limit {
+            // Safe to unwrap, the priority queue cannot be empty.
+            let (memory_usage_before, canister_id) = priority_queue.pop_last().unwrap();
```

Comment:
Is this really true? If there are no canisters or no messages within any of the canisters, then the priority queue would be empty. 

So it seems to me that this `unwrap` is only safe for any `limit` (in particular `0`) if we are sure that `memory_usage` is also `0` in the cases above.

We could simply do 
```suggestion
            let (memory_usage_before, canister_id) = priority_queue.pop_last() else { return };
```
to avoid having to reason about this?

Reply:
> If there are no canisters or no messages within any of the canisters, then the priority queue would be empty.

We also compute `memory_usage` above, while building the priority queue. In such a way that it's equal to the sum of all enqueued canister's memory usages. Then below, we update `memory_usage` every time we shed (or fail to shed) a message. So unless `best_effort_message_memory_usage()` is non-deterministic (or the canister state is updated concurrently, which is kind of the same), the priority queue can only be empty iff `memory_usage == 0`.

IOW, unwrapping here is always safe. However, as per the comment thread above, if `CanisterQueues::best_effort_message_memory_usage()` starts returning non-zero values for canisters with zero best-effort messages (or at least zero messages that can be shed), then we may end up in an infinite loop (because we keep calling `shed_largest_message()`, which is a no-op in that situation; and replacing the offending canister into the priority queue). I'll try to rewrite the code to work around that issue, however unlikely it is.